Selectmenu: Add classes option

gabrielschulhof opened the pull request.

Commit:
Selectmenu: Implement classes option

Fixes gh-7731

arschmitz commented:
@gabrielschulhof were you wanting a review of this

gabrielschulhof commented:
@arschmitz I've made a few more changes and I could use a review although I'm fairly certain it's not ready to go yet, but it'll be good to have another set of eyes on it so I have a clear list of what remains to be done. TIA!

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -79,16 +94,24 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
 		if ( event.type === "vclick" ||
-				event.keyCode && ( event.keyCode === $.mobile.keyCode.ENTER || event.keyCode === $.mobile.keyCode.SPACE ) ) {
+				event.keyCode &&
+					( event.keyCode === $.mobile.keyCode.ENTER ||
```
use `$.ui.keyCode` this is deprecated

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -79,16 +94,24 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
 			if ( this.menuType === "overlay" ) {
-				this.button.attr( "href", "#" + this.popupId ).attr( "data-" + ( $.mobile.ns || "" ) + "rel", "popup" );
+				this.button
+					.attr( "href", "#" + this.popupId )
```
is popupId escaped elsewhere?

gabrielschulhof replied:
No, and it doesn't need to be, because I'm not constructing a selector with it.

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -79,16 +94,24 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-				this.button.attr( "href", "#" + this.popupId ).attr( "data-" + ( $.mobile.ns || "" ) + "rel", "popup" );
+				this.button
+					.attr( "href", "#" + this.popupId )
+					.attr( "data-" + ( $.mobile.ns || "" ) + "rel", "popup" );
```
ns is deprecated move determining this to backcompat and make an extension point

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -79,16 +94,24 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
 			} else {
-				this.button.attr( "href", "#" + this.dialogId ).attr( "data-" + ( $.mobile.ns || "" ) + "rel", "dialog" );
+				this.button
+					.attr( "href", "#" + this.dialogId )
```
same as above is this escaped

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -79,16 +94,24 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-				this.button.attr( "href", "#" + this.dialogId ).attr( "data-" + ( $.mobile.ns || "" ) + "rel", "dialog" );
+				this.button
+					.attr( "href", "#" + this.dialogId )
+					.attr( "data-" + ( $.mobile.ns || "" ) + "rel", "dialog" );
```
move to backcompat

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -233,33 +257,40 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-			"<div class='ui-title'></div>" +
-			"</div>" +
-			"<div class='ui-content'></div>" +
+		menuPage = $( "<div data-" + $.mobile.ns + "role='dialog'>" +
```
move to backcompat

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -51,14 +53,31 @@ var unfocusableItemSelector = ".ui-disabled,.ui-state-disabled,.ui-listview-item
 		// a parent is a popup
-		o.nativeMenu = o.nativeMenu || ( this.element.parents( ":jqmData(role='popup'),:mobile-popup" ).length > 0 );
+		o.nativeMenu = o.nativeMenu ||
+			( this.element.parents( ":jqmData(role='popup'),:mobile-popup" ).length > 0 );
```
dont use jqmdata its deprecated.

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -420,57 +461,91 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-		var self = this,
-			$window = this.window,
-			selfListParent = self.list.parent(),
+		var $window = this.window,
```
remove hungarian notation

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -420,57 +461,91 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
 
 			// For WebOS/Opera Mini (set lastscroll using button offset)
 			if ( scrollTop === 0 && buttonOffset > screenHeight ) {
-				self.thisPage.one( "pagehide", function() {
```
pagehide is deprecated

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -420,57 +461,91 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-			self.menuPage.one( {
+			this.menuPage.one( {
 				pageshow: $.proxy( this, "_focusMenuItem" ),
 				pagehide: $.proxy( this, "close" )
```
use page container events

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -523,7 +598,10 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
+			if ( needPlaceholder &&
+				( !option.getAttribute( "value" ) ||
+					text.length === 0 ||
+					$option.jqmData( "placeholder" ) ) ) {
```
dont use jqmdata

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -320,4 +334,6 @@ return $.widget( "mobile.selectmenu", $.extend( {
 	}
 }, $.mobile.behaviors.formReset ) );
```
use UI form reset and also $.widget now supports mixins via passing an array

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -78,15 +82,16 @@ return $.widget( "mobile.selectmenu", $.extend( {
+		// suffix. See https://github.com/jquery/jquery-mobile/issues/3577
+		// if ( $el[0].className.length ) {
+		// 	classes = $el[0].className;
+		// }
```
this is no longer relavent

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -95,49 +100,54 @@ return $.widget( "mobile.selectmenu", $.extend( {
-			classes += " ui-mini";
-		}
+		this._removeClass( this.element, null, "ui-button-left ui-button-right" );
+		this.select = this.element;
```
whats the point of `this.select` if its always `this.element`?

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -95,49 +100,54 @@ return $.widget( "mobile.selectmenu", $.extend( {
 	_create: function() {
+		var options = this.options,
+			iconpos = options.icon ?
+				( options.iconpos || this.select.jqmData( "iconpos" ) ) : false;
```
dont use `jqmData`

arschmitz commented:
Looking good done for now

Commit:
Dependencies: add jquery-ui/form-reset-mixin.js

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -51,14 +53,32 @@ var unfocusableItemSelector = ".ui-disabled,.ui-state-disabled,.ui-listview-item
+	options: {
+		classes: {
+			"ui-selectmenu-custom-header-close-button":
+				"ui-button ui-corner-all ui-button-left ui-button-icon-only ui-icon-delete"
```
`ui-button*` is not a theme class neither is `ui-icon*`

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -227,39 +256,46 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
+			.prependTo( listbox );
+		headerTitle = $( "<h1></h1>" ).appendTo( header );
+
+		this._addClass( menuPage, null, "ui-selectmenu-custom" );
```
`ui-selectmenu*` should be in the second param they belong to this widget

arschmitz commented on js/widgets/forms/select.custom.js:
```diff
@@ -227,39 +256,46 @@ return $.widget( "mobile.selectmenu", $.mobile.selectmenu, {
-			} ).appendTo( header );
+				"href": "#"
+			} );
+			this._addClass( headerClose, "ui-selectmenu-custom-header-close-button" );
```
same here

gabrielschulhof replied:
I don't get it. ```ui-selectmenu-custom-header-close-button``` *is* in the second param here.

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -95,49 +94,55 @@ return $.widget( "mobile.selectmenu", $.extend( {
-				this.element.addClass( wrapper.hasClass( "ui-button-left" ) ? "ui-button-left" : "ui-button-right" );
+		if ( this.selectWrapper.length > 0 ) {
+			if ( this.selectWrapper.is( ".ui-button-left, .ui-button-right" ) ) {
+				this._addClass( null,
```
button classes should be in the thuird param

arschmitz replied:
also no need to remove classes in destroy

arschmitz replied:
no need to do anything you remove the whole element below

gabrielschulhof replied:
This adds classes back to ```this.element``` upon destroy.

gabrielschulhof replied:
... and if you leave out the element to which you wish to apply the classes, it becomes ```this.element```, so the classes are in the correct parameter, AFAICT.

gabrielschulhof replied:
Wait a sec - if I use ```this._addClass()``` here, classes will be added but then immediately removed, right? Hmmm ...

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -95,49 +94,55 @@ return $.widget( "mobile.selectmenu", $.extend( {
 	_create: function() {
+		var options = this.options,
+			iconpos = options.icon ?
+				( options.iconpos || this.element.jqmData( "iconpos" ) ) : false;
```
dont use `jqmdata`

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -214,7 +221,7 @@ return $.widget( "mobile.selectmenu", $.extend( {
 			}
 		} );
-		self.select.bind( "focus", function() {
+		self.element.bind( "focus", function() {
```
should really avoid self since in some contexts in can reference window

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -266,16 +274,18 @@ return $.widget( "mobile.selectmenu", $.extend( {
 				// TODO possibly aggregate multiple select option classes
 				return span
-					.addClass( self.select.attr( "class" ) )
+					.addClass( self.element.attr( "class" ) )
```
using addClass still

arschmitz commented:
looking good only a few things

gabrielschulhof commented:
@jaspermdegroot can you please check http://view.jquerymobile.com/7731-selectmenu-classes-option/demos/selectmenu-custom/ and tell me if I'm doing anything wrong with the classes? The styling is obviously wrong, but is it wrong because I'm adding/failing to add classes wrong, or is it wrong because the contents of the CSS is wrong?

gabrielschulhof commented:
@jaspermdegroot I have now pushed the commit that removes the class juggling.

http://view.jquerymobile.com/7731-selectmenu-classes-option/demos/selectmenu-custom/

gabrielschulhof commented:
@jaspermdegroot reminder: 9743d32a56aaef242d29ec184495bed68e74c50c removes the fix for https://github.com/jquery/jquery-mobile/issues/5073, but you said it should be fixed in CSS.

jaspermdegroot commented:
@gabrielschulhof @arschmitz 

This test page shows the issue from #5073: http://jsbin.com/zikohihiyu/edit?html,output

I can add CSS to negate the rules from `ui-toolbar-header-button-left/right` when the class is on a span inside a custom select menu, but then the user still doesn't get the header left/right button positioning that (s)he wants. When that ticket was opened we were automatically adding the left/right button classes but we don't do that anymore, so I think it's best to document / demonstrate that in case of a select you have to wrap it in a div and add the button left/right class to that div, never on the `select` itself.

What did we decide about how we handle classes in case of generated wrappers? To add a class to the container you need to use the classes option (or enhance the selectmenu yourself once we added that option)? Adding a class to the `select` itself is not supported?

arschmitz commented:
@jaspermdegroot classes option would be the only fuly supported way to do that

Commit:
Page: Use _.on() for pagehide

Commit:
Selectmenu: Update tests to use bootstrap

Commit:
Selectmenu: Fix assumptions in option test

gabrielschulhof commented:
@jaspermdegroot could you please have a look at the link below?

http://jsbin.com/jujobam/1/

It is based on this PR.

Even though the classes ```ui-mini``` and ```ui-toolbar-header-button-right``` are getting applied to the selectmenu wrapper, ```position: absolute;``` from ```ui-toolbar-header-button-right``` is getting overridden by ```position: relative;``` from ```ui-selectmenu``` because the latter seems to have precedence.

What is the best way to resolve this?

There also seems to be a problem with the arrow icon in the button. Can you tell me what I'm doing wrong?

jaspermdegroot commented:
@gabrielschulhof 

> Even though the classes ui-mini and ui-toolbar-header-button-right are getting applied to the selectmenu wrapper, position: absolute; from ui-toolbar-header-button-right is getting overridden by position: relative; from ui-selectmenu because the latter seems to have precedence.

I think we can remove `position: relative;` for `ui-selectmenu`. The `select` element has `position: absolute;` so one of the wrapper `div`s should get `position: relative;` (or anything other than the default `static`). `ui-button` already has `relative` so I think it's safe to remove it from the outer wrapper.

Have to check with a custom selectmenu as well. I can't preview this branch on view.* since it's in your fork. Just see if anything breaks if you remove it. I can always double check after this has been merged.

> There also seems to be a problem with the arrow icon in the button. Can you tell me what I'm doing wrong?

I suppose you mean the lack of space between text and icon? Since we make the icon float with class `ui-widget-icon-floatend` we will have to add extra padding to the button. Maybe we should do that in core/button CSS, not in selectmenu CSS. Will look into that.

Commit:
Selectmenu: Remove relative positioning

As per https://github.com/jquery/jquery-mobile/pull/8309#issuecomment-206440536

gabrielschulhof commented:
@jaspermdegroot 

Thanks for looking into it! I've removed ```position: relative;``` from ```ui-selectmenu``` and it's all good now.

Meanwhile, here's a bin illustrating both custom and non-custom select in both a toolbar and non-toolbar setting. The bin should look right once view updates the branch.

http://jsbin.com/zukase/1

I've pushed this branch upstream as well. http://view.jquerymobile.com/7731-selectmenu-classes-option/demos/

gabrielschulhof commented:
@jaspermdegroot the problem with the arrow icon on the selectmenu is actually pretty bad on Firefox:

![7731-firefox](https://cloud.githubusercontent.com/assets/976081/14344802/6ee0d5f8-fcb2-11e5-834f-16c072e67752.png)

Commit:
Selectmenu: Icon span must precede text span

gabrielschulhof commented:
OK, so if the icon span precedes the label span then FF looks OK. Probably because we never set the top and it's absolutely positioned and in FF, the "cursor" wraps to the next line, whereas in Chrome it does not.

gabrielschulhof commented:
http://output.jsbin.com/zukase/ fetches the CSS from rawgit.com, so that shows the updated CSS, but unfortunately, I can't show the updated JS because only view can give me that service, and view doesn't seem to be updating.

Commit:
Selectmenu: Add class key to button text span

gabrielschulhof commented:
Yaaay! view has updated! http://jsbin.com/zukase/1 now shows both the removal of ```position: absolute;``` and of having rearranged the spans such that the icon comes first. Works in both FF and Chrome.

gabrielschulhof commented:
I think this is ripe for another round of review.

Commit:
Selectmenu: Update toolbar-related class names

Commit:
Selectmenu: Fix popup header close button style

jaspermdegroot commented:
The lack of space between text and icon should be fixed with this PR: https://github.com/jquery/jquery-mobile/pull/8426

jaspermdegroot commented:
CSS looks good. There are problems with the custom selectmenu popup on FF though.

Commit:
Selectmenu: Create toolbars late to make sure they find containing page

Commit:
Core: Add ui- namespace to page selectors

arschmitz commented on external/qunit-assert-classes/qunit-assert-classes.js:
```diff
@@ -58,8 +58,8 @@
-			missing: missing,
-			found: found,
+			missing: missing || [],
+			found: found || [],
```
this cant be changed directly this is an external dependency

arschmitz commented on js/widgets/forms/select.js:
```diff
@@ -25,119 +25,110 @@
 } )( function( $ ) {
 
-return $.widget( "mobile.selectmenu", $.extend( {
+return $.widget( "mobile.selectmenu", $.widget( "mobile.selectmenu", [ {
```
this call is weird i think it would be more clear to separate out adding the theme

arschmitz commented:
Looks mostly good just those 2 things plus what @jaspermdegroot mentioned

Commit:
Selectmenu: Address review comments

Commit:
Widget: Define $.mobile.widget.backcompat in all cases

$.mobile.widget.backcompat must not be undefined even if $.mobileBackcompat is
false, since other widgets unconditionally extend with
$.mobile.widget.backcompat and if $.mobile.widget.backcompat is undefined, then
they end up doing $.widget( ..., undefined ) which results in _create() being
assigned $.noop.

Commit:
Selectmenu: Remove list from dialog only after it is completely hidden

Commit:
Button: Correct spelling error

Commit:
Selectmenu: Fix controlgroup interaction

Commit:
Controlgroup: Fix themeing

Commit:
All: Address review comments

Commit:
Tests: Add define() wrappers

Commit:
All: Remove UI core and references to it

gabrielschulhof commented:
@arschmitz we have a load order race condition in kitchensink which seems to reliably manifest itself with core 1.9.1 in this PR. Aside from that it should be golden. I'll try to look into it whenever I get a chance.

Commit:
Tests: Use $.Deferred() in kitchensink

gabrielschulhof commented:
@arschmitz nm, using a ```$.Deferred()```. Lessee...

gabrielschulhof commented:
Yaaaay! Tests are now no worse than 1.5-dev!

arschmitz commented:
👍

apsdehal commented:
Awesome! Landed in 1.5-dev.
